testsuite/gateways/apicast/template.py

```diff
@@ -1,156 +1,107 @@
 """Self managed apicast deployed from apicast template"""
 import logging
-from abc import ABC, abstractmethod
-from typing import Optional, List
-from urllib.parse import urlparse
+import os
 
-from threescale_api.resources import Service
+import importlib_resources as resources
 
-from testsuite.openshift.objects import Routes, SecretTypes
-from testsuite.requirements import ThreeScaleAuthDetails
-from .selfmanaged import SelfManagedApicast, SelfManagedApicastRequirements
-from ...openshift.env import Environ
+from testsuite.openshift.objects import SecretTypes
+from .external import ExternalApicast
+from ...openshift.client import OpenShiftClient
 
 LOGGER = logging.getLogger(__name__)
 
 
-class TemplateApicastRequirements(SelfManagedApicastRequirements,
-                                  ThreeScaleAuthDetails,
-                                  ABC):
-    """Requirements for TemplateApicast"""
-    @property
-    @abstractmethod
-    def template(self):
-        """Returns template file"""
-
-    @property
-    @abstractmethod
-    def image(self):
-        """Return image to use with in the template"""
-
-    @property
-    @abstractmethod
-    def configuration_url(self):
-        """Returns URL for configuring apicast"""
-
-    @property
-    @abstractmethod
-    def service_routes(self) -> bool:
-        """True, if apicast should creates route for every service"""
-
-
-# pylint: disable=too-many-instance-attributes
-class TemplateApicast(SelfManagedApicast):
-    """Template-based Apicast Gateway."""
-
-    def __init__(self, requirements: TemplateApicastRequirements) -> None:
-        super().__init__(requirements)
-        self.requirements = requirements
-        self.template = requirements.template
-        self.image = requirements.image
-        self.service_routes = requirements.service_routes
-        self.service_name = self.deployment
-        self.name = self.deployment
-        self.configuration_url_secret_name = f'{self.deployment}-config-url'
-        self.routes: List[str] = []
-        self.route_type = Routes.Types.EDGE
-
-    def get_app_params(self, **kwargs):
-        """Template envs for oc new-app."""
-        params = {
-            "APICAST_NAME": self.deployment,
-            "AMP_APICAST_IMAGE": self.image,
-            "DEPLOYMENT_ENVIRONMENT": "production",
-            "CONFIGURATION_LOADER": "boot",
-            "CONFIGURATION_CACHE": 300,
-            "LOG_LEVEL": "info",
-            "CONFIGURATION_URL_SECRET": self.configuration_url_secret_name,
-        }
+class TemplateApicast(ExternalApicast):
```

Comment:
now I am confused. `ExternalApicast` is presented as already deployed apicast, this not the case of `TemplateApicast`.
??

Reply:
`ExternalApicast` expects the Apicast to be already deployed, however, the `TemplateApicast` and `OperatorApicast` extend this behaviour and add the implementation for creation/deletion.

Reply:
then we are back at the beginning, it is neither `ExternalApicast` not `IndependentApicast`, it is `SelfmanagedApicast`